README.md
Blackarch Install
-----------------

Install From ArchLinux/BlackArch Live-ISO-DVD/BlackArch NetInstall
----------------------------------------------------------------
Version: 0.5.x (See Github Tags)

    1. # blackarch-install
    2. Select Install Mode:
        1. Install from Live-ISO.
            - It will dump the current Live image into your hard drive.
        2. Install from BlackArch Official Repository.
            - It will install base system from Arch Linux and
            add Black Arch Linux packages from the official repository
            into your hard drive. You can choose the packets to install.
        3. Install from Blackman.
            - It will install base system from Arch Linux and
            will download and compile all the BlackArch Linux packages
            from their source according to PKGBULD files.
    3. Partitioning:
        3.1. Set the device you want to install BlackArch Linux in.
            -> i.e. /dev/sda (default one)
        3.2. Choose Partition Mode:
            * At the moment, only /boot and / root partition are supported.
            * We will add optional /home and Swap in future releases.
            1. Automatic - only empty HD! new partitions will be created in
               /dev/sda1 and /dev/sda2.
                - This mode its thinking for Virtual Machines environment,
                where hard disk isnt contain any partitions already.
                It will create by default sda1 for /boot partition
                and sda2 for / root partition. Also, it will format
                /boot partition with ext2 type and / root partition with
                ext4 type.
                - The only configurable thing here its the size of both
                partitions. Set it by default to 100MB for /boot and
                10GB for / root partition.
                - WARNING: this mode isnt tested properly in non Virtual
                Machine box. So do not used it at least you know what
                you are doing. You had been warned bitch. ;)
            2. Manual - set manual partitions with cfdisk and format them.
                - This mode will run cfdisk utility for partition creation.
                Once you have set up your proper /boot and / root partition,
                dont forget to Write and you are good to go.
                It will ask you when Quit for the /boot and / root partition
                number chosed, also the partition type is required for
                format purposes.
    4. Install
        - BlackArch Linux will be now installed into your hard drive.
        This will take a while.
    5. Setup up root password.
    6. Setup normal user account (optional).
    7. Grub install (automatic)
    8. Reboot and enjoy your Black Arch distro.


Install From ArchLinux - v0.4 Release (See Github Tags)
--------------------------------------------------------
    - v0.4 comes with several install options:
        1. Install from Live-ISO - Dump everything to your partition.
        2. Install from Repository - Install packages you choose From BlackArch Official Repository.
        3. Install from Blackman - Download And Compile Packages from their PKGBUILD.


Install From ArchLinux - v0.1 Release (See Github Tags)
-------------------------------------------------------
   1. Download ISO - https://www.archlinux.org/download/
	-> i.e. http://archlinux.c3sl.ufpr.br/iso/2013.12.01/archlinux-2013.12.01-dual.iso
   2. Create Vbox or VMWare Box for it - Do not use this script in real ENV.
   3. # pacman -Syyu
   4. # pacman -S git
   5. git clone https://github.com/BlackArch/blackarch-install-scripts
   6. cd blackarch-install-scripts; sed 's/SEC_ENABLE="true"/SEC_ENABLE="false"/' blackarch-install; sh blackarch-install
   7. Enjoy BlackArch!


changelog
v0.1 - First release
	-> install from Official BlackArch Repositories
v0.2
	-> splash screen for grub
	-> Install
		2. Install from Repository - Install packages you choose From
			BlackArch Official Repository.
		3. Install from Blackman - Download And Compile
			Packages from their PKGBUILD.
v0.3
	-> Partition
		-> added Automatic and Manual Mode Menu
			-> Manual not implemented (v0.5)
v0.4
	-> install
		1. Install from Live-ISO - Dump everything to your partition.
v0.5
	-> Partition mode
		-> Manual mode implemented


chroot-install.sh
#!/bin/bash
################################################################################
#                                                                              #
# blackarchinstall - Blackarch Install                                         #
#                                                                              #
# FILE                                                                         #
# chroot-install.sh                                                            #
#                                                                              #
# DATE                                                                         #
# 2013-12-16                                                                   #
#                                                                              #
# DESCRIPTION                                                                  #
# Script for easy install                                                      #
#                                                                              #
# AUTHOR                                                                       #
# nrz@nullsecurity.net                                                         #
#                                                                              #
################################################################################

# HD beta var
HD="sda"

# blackarchinstall version
VERSION="v0.2"

# true / false
FALSE="0"
TRUE="1"

# return codes
SUCCESS="1337"
FAILURE="31337"

# verbose mode - default: quiet
VERBOSE="/dev/null"

# colors
WHITE="$(tput bold ; tput setaf 7)"
GREEN="$(tput setaf 2)"
RED="$(tput bold; tput setaf 1)"
YELLOW="$(tput bold ; tput setaf 3)"
NC="$(tput sgr0)" # No Color

# BA REPO
BLACKARCH_REPO_URL='http://www.blackarch.org/blackarch/$repo/os/$arch'

wprintf() {
    fmt=$1
    shift
    printf "%s${fmt}%s\n" "${WHITE}" "$@" "${NC}"

    return "${SUCCESS}"
}

# print warning
warn()
{
    printf "%s[!] WARNING: %s%s\n" "${RED}" "${*}" "${NC}"

    return "${SUCCESS}"
}

# print error and exit
err()
{
    printf "%s[-] ERROR: %s%s\n" "${RED}" "${*}" "${NC}"

    return "${SUCCESS}"
}

# print error and exit
cri()
{
    printf "%s[-] CRITICAL: %s%s\n" "${RED}" "${*}" "${NC}"

    exit "${FAILURE}"
}


# usage and help
usage()
{
cat <<EOF
Usage: $0 <arg> | <misc>
OPTIONS:
    -i: install
MISC:
    -V: print version and exit
    -H: print help and exit
EOF
    return "${SUCCESS}"
}

check_env()
{
    if [ -f /var/lib/pacman/db.lck ]; then
        cri "Pacman locked - rm /var/lib/pacman/db.lck"
    fi
}

# check argument count
check_argc()
{
    return "${SUCCESS}"
}

# check if required arguments were selected
check_args()
{
    return "${SUCCESS}"
}

update_system()
{
    if ! grep -q "blackarch" /etc/pacman.conf; then
        wprintf "[+] Adding BlackArch Official Repo"
        printf '[blackarch]\nServer = %s\n' "${BLACKARCH_REPO_URL}" >> /etc/pacman.conf
    fi

    # key problem - will be solved later on
    pacman-key --init
    pacman -Syyu --noconfirm
    pacman-key -r 4345771566D76038C7FEB43863EC0ADBEA87E4E3
    pacman-key --lsign-key 4345771566D76038C7FEB43863EC0ADBEA87E4E3
    pacman -Syy

    return "${SUCCESS}"
}

install_packages()
{
    pacman -S blackarch
    return "${SUCCESS}"
}

install_grub()
{
    pacman -S grub --noconfirm

    wprintf "[+] Patching grub2 syntax error"
    printf "GRUB_DISABLE_SUBMENU=y\n" >> /etc/default/grub

    grub-install "/dev/${HD}"
    grub-mkconfig -o /boot/grub/grub.cfg

    return "${SUCCESS}"
}

install()
{
    wprintf "[+] Updating system..."
    update_system

    wprintf "[+] Installing packages..."
    install_packages

    wprintf "[+] Installing grub..."
    install_grub

    return "${SUCCESS}"
}

# parse command line options
get_opts()
{
    while getopts ivVH flags
    do
        case "${flags}" in
            i)
                #optarg=${OPTARG}
                opt="install"
                ;;
            v)
                VERBOSE="/dev/stdout"
                ;;
            V)
                printf "%s\n" "${VERSION}"
                exit "${SUCCESS}"
                ;;
            H)
                usage
                ;;
            *)
                err "WTF?! mount /dev/brain"
                ;;
        esac
    done

    return "${SUCCESS}"
}


# controller and program flow
main()
{
    check_argc ${*}
    get_opts ${*}
    check_args ${*}
    check_env

    # commented arg opt
    install

    return "${SUCCESS}"
}


# program start
main ${*}

# EOF


blackarchinstall.sh
#!/bin/bash
################################################################################
#                                                                              #
# blackarchinstall - Blackarch Install                                         #
#                                                                              #
# FILE                                                                         #
# blackarchinstall.sh                                                          #
#                                                                              #
# DATE                                                                         #
# 2013-12-16                                                                   #
#                                                                              #
# DESCRIPTION                                                                  #
# Script for easy install                                                      #
#                                                                              #
# AUTHOR                                                                       #
# nrz@nullsecurity.net                                                         #
#                                                                              #
################################################################################

# SECURITY VAR - this version can rm -rf /* your hard drive
SEC_ENABLE="false"

# HD beta var
HD="sda"

# blackarchinstall version
VERSION="blackarchinstall v0.1"

# true / false
FALSE="0"
TRUE="1"

# return codes
SUCCESS="1337"
FAILURE="31337"

# verbose mode - default: quiet
VERBOSE="/dev/null"

# colors
WHITE="$(tput bold ; tput setaf 7)"
GREEN="$(tput setaf 2)"
RED="$(tput bold; tput setaf 1)"
YELLOW="$(tput bold ; tput setaf 3)"
NC="$(tput sgr0)" # No Color

# root variable
ROOT="/mnt/blackarch"


gprintf() {
    fmt=$1
    shift
    printf "%s${fmt}%s\n" "${WHITE}" "$@" "${NC}"

    return "${SUCCESS}"
}

# print warning
warn()
{
    printf "%s[!] WARNING: %s%s\n" "${RED}" "${*}" "${NC}"

    return "${SUCCESS}"
}

# print error and exit
err()
{
    printf "%s[-] ERROR: %s%s\n" "${RED}" "${*}" "${NC}"

    return "${SUCCESS}"
}

# print error and exit
cri()
{
    printf "%s[-] CRITICAL: %s%s\n" "${RED}" "${*}" "${NC}"
    
    exit "${FAILURE}"
}


# usage and help
usage()
{
cat <<EOF
Usage: $0 <arg> | <misc>
OPTIONS:
    -i: install
MISC:
    -V: print version and exit
    -H: print help and exit
EOF
    return "${SUCCESS}"
}

# leet banner, very important
banner()
{
    printf "%s--==[ BlackArch Install ]==--%s\n" "${YELLOW}" "${NC}"

    return "${SUCCESS}"
}

check_env()
{
    if [ -f /var/lib/pacman/db.lck ]; then
        cri "Pacman locked - rm /var/lib/pacman/db.lck"
    fi
}

# check argument count
check_argc()
{
    return "${SUCCESS}"
}

# check if required arguments were selected
check_args()
{
    return "${SUCCESS}"
}


update_system()
{
    if ! grep -q "blackarch" /etc/pacman.conf; then
        gprintf "[+] Adding BlackArch Official Repo"
        printf '[blackarch]\nServer = http://www.blackarch.org/pub/blackarch/$arch\n' >> /etc/pacman.conf
    fi
    gprintf "[+] Enabling multilib"
    printf "[multilib]\nInclude = /etc/pacman.d/mirrorlist\n" >> /etc/pacman.conf

    #fix this
    #sed -i '/[multilib]/{ s/^#//; n; s/^#//; }' /etc/pacman.conf

    # key problem - will be solved later on
    pacman -Syyu --noconfirm
    pacman-key --lsign-key 4345771566D76038C7FEB43863EC0ADBEA87E4E3
    pacman -Syy
    
    return "${SUCCESS}"
}

format_filesystem()
{
    # TODO: add swap 

    parted -s "/dev/${HD}" mklabel msdos

    # boot partition
    parted -a optimal "/dev/${HD}" mkpart primary 0% 100MB
    # partition bootable
    parted "/dev/${HD}" set 1 boot on
    # root partition
    parted -a optimal "/dev/${HD}" mkpart primary 100MB 100%

    # formating
    mkfs.ext2 -L boot /dev/sda1
    mkfs.ext4 -L root /dev/sda2

    gprintf "[+] Created boot partition: /dev/sda1 - ext2"
    gprintf "[+] Created root partition: /dev/sda2 - ext4"

    
    return "${SUCCESS}"
}

mount_filesystem()
{
    #TODO: change boot device for user required
    
    mkdir -p "${ROOT}"
    mount -t ext4 "/dev/${HD}2" "${ROOT}" 
    mkdir -p "${ROOT}/boot"
    mount -t ext2 "/dev/${HD}1" "${ROOT}/boot"
      
    return "${SUCCESS}"
}

install_packages()
{
    pacstrap -c "${ROOT}" base 
    pacstrap -c "${ROOT}" blackarch

    return "${SUCCESS}"
}

install_grub()
{
    # patching grub2 syntax error
    echo "GRUB_DISABLE_SUBMENU=y" >> "${ROOT}"/etc/default/grub

    cp grub-install.sh "${ROOT}/bin"
    arch-chroot "${ROOT}" grub-install.sh
    
    # cleaning up
    rm -rf "${ROOT}/bin/grub-install.sh"

    return "${SUCCESS}"
}

install()
{
    gprintf "[+] Updating system..."
    update_system
    
    gprintf "[+] Partitioning ${HD}..."
    format_filesystem

    gprintf "[+] Mounting filesystem..."
    mount_filesystem

    gprintf "[+] Installing BlackArch packages..."
    install_packages
    
    gprintf "[+] Generating fstab..."
    genfstab -p "${ROOT}" >> "${ROOT}/etc/fstab"

    gprintf "[+] Installing grub..."
    install_grub

    return "${SUCCESS}"
}

# parse command line options
get_opts()
{
    while getopts ivVH flags
    do
        case "${flags}" in
            i)
                #optarg=${OPTARG}
                opt="install"
                ;;
            v)
                VERBOSE="/dev/stdout"
                ;;
            V)
                printf "%s\n" "${VERSION}"
                exit "${SUCCESS}"
                ;;
            H)
                usage
                ;;
            *)
                err "WTF?! mount /dev/brain"
                ;;
        esac
    done

    return "${SUCCESS}"
}


# controller and program flow
main()
{
    banner
    check_argc ${*}
    get_opts ${*}
    check_args ${*}
    check_env

    if [ "${opt}" == "install" ]; then
        if [[ "${SEC_ENABLE}" == "false" ]]; then
            install
        else
            cri "SEC_ENABLE var active"
        fi
    fi
    return "${SUCCESS}"
}


# program start
main ${*}

# EOF
